File: typescript_bun/bowling_game/bowling_game.test.ts
import { describe, beforeEach, expect, test } from "bun:test";
import { BowlingGame } from "./bowling_game";

describe("Bowling Game", () => {
    let bg: BowlingGame;

    beforeEach(() => {
        bg = new BowlingGame();
    });

    function rollMany(game: BowlingGame, times: number, pins: number) {
        for (let i = 0; i < times; i++) {
            game.roll(pins);
        }
    }

    test("All zeros game", () => {
        rollMany(bg, 20, 0);
        expect(bg.score()).toBe(0);
    });

    test("All one game", () => {
        rollMany(bg, 20, 1);
        expect(bg.score()).toBe(20);
    });

    test("One spare game", () => {
        bg.roll(5);
        bg.roll(5);
        bg.roll(3);
        rollMany(bg, 17, 0);
        expect(bg.score()).toBe(16);
    });

    test("One strike game", () => {
        bg.roll(10);
        bg.roll(3);
        bg.roll(4);
        rollMany(bg, 16, 0);
        expect(bg.score()).toBe(24);
    });

    test("Perfect game", () => {
        rollMany(bg, 12, 10);
        expect(bg.score()).toBe(300);
    });
});

File: zig/src/lib/bowling_game_test.zig
const std = @import("std");
const BowlingGame = @import("bowling_game.zig").BowlingGame;

test "All zero game" {
    var bg = BowlingGame.init();
    for (0..20) |_| {
        bg.roll(0);
    }
    try std.testing.expect(bg.score() == 0);
}

test "All ones game" {
    var bg = BowlingGame.init();
    for (0..20) |_| {
        bg.roll(1);
    }
    try std.testing.expect(bg.score() == 20);
}

test "one spare game" {
    var bg = BowlingGame.init();
    bg.roll(5);
    bg.roll(5);
    bg.roll(3);
    for (0..17) |_| {
        bg.roll(0);
    }
    try std.testing.expect(bg.score() == 16);
}

test "one strike game" {
    var bg = BowlingGame.init();
    bg.roll(10);
    bg.roll(3);
    bg.roll(4);
    for (0..16) |_| {
        bg.roll(0);
    }
    try std.testing.expect(bg.score() == 24);
}

test "perfect game" {
    var bg = BowlingGame.init();
    for (0..12) |_| {
        bg.roll(10);
    }
    try std.testing.expect(bg.score() == 300);
}
